compiler/luci-interpreter/src/kernels/Add.test.cpp

```diff
@@ -27,59 +27,118 @@ namespace
 
 using namespace testing;
 
-TEST(AddTest, Float)
+// for quantized Add, the error shouldn't exceed step
+float GetTolerance(float min, float max)
 {
-  Shape base_shape = {2, 3, 1, 2};
-  std::vector<Shape> test_shapes{{1, 1, 3, 2}, {1, 3, 1, 2}, {2, 1, 3, 1}, {2, 3, 1, 1}};
-  std::vector<std::vector<float>> test_outputs = {
-      {0.0f, 2.6f, 0.0f, 2.8f, 0.7f, 3.2f, 1.1f, 0.8f, 0.5f, 1.0f, 1.9f, 1.4f,
-       1.0f, 0.0f, 0.4f, 0.0f, 1.8f, 0.0f, 1.4f, 3.1f, 0.8f, 3.3f, 2.2f, 3.7f,
-       0.0f, 0.3f, 0.0f, 0.5f, 0.0f, 0.9f, 0.9f, 0.0f, 0.3f, 0.0f, 1.7f, 0.0f},
-      {0.0f, 2.6f, 0.5f, 1.0f, 1.8f, 0.0f, 1.4f, 3.1f, 0.0f, 0.5f, 1.7f, 0.0f},
-      {0.0f, 2.5f, 0.0f, 2.6f, 0.0f, 1.9f, 1.1f, 0.7f, 1.2f, 0.8f, 0.5f, 0.1f,
-       1.0f, 0.0f, 1.1f, 0.0f, 0.4f, 0.0f, 1.7f, 3.3f, 2.2f, 3.8f, 2.1f, 3.7f,
-       0.0f, 0.5f, 0.0f, 1.0f, 0.0f, 0.9f, 1.2f, 0.0f, 1.7f, 0.0f, 1.6f, 0.0f},
-      {0.0f, 2.5f, 1.2f, 0.8f, 0.4f, 0.0f, 1.7f, 3.3f, 0.0f, 1.0f, 1.6f, 0.0f}};
-  std::vector<float> input1_data{-0.3f, 2.3f, 0.9f,  0.5f, 0.8f, -1.1f,
-                                 1.2f,  2.8f, -1.6f, 0.0f, 0.7f, -2.2f};
-  std::vector<float> input2_data{0.2f, 0.3f, -0.4f, 0.5f, 1.0f, 0.9f};
-  for (size_t i = 0; i < test_shapes.size(); ++i)
-  {
-    Tensor input1_tensor = makeInputTensor<DataType::FLOAT32>(base_shape, input1_data);
-    Tensor input2_tensor = makeInputTensor<DataType::FLOAT32>(test_shapes[i], input2_data);
-    Tensor output_tensor = makeOutputTensor(DataType::FLOAT32);
+  float kQuantizedStep = (max - min) / 255.0;
+  return kQuantizedStep;
+}
 
-    AddParams params{};
-    params.activation = Activation::RELU;
+float getMinValue(std::initializer_list<float> data)
+{
+  return 0 > std::min<float>(data) ? std::min<float>(data) : 0;
+}
 
-    Add kernel(&input1_tensor, &input2_tensor, &output_tensor, params);
-    kernel.configure();
-    kernel.execute();
+float getMaxValue(std::initializer_list<float> data)
+{
+  return 0 < std::max<float>(data) ? std::max<float>(data) : 0;
+}
 
-    EXPECT_THAT(extractTensorData<float>(output_tensor),
-                ::testing::ElementsAreArray(ArrayFloatNear(test_outputs[i], 0.0001f)))
-        << "With shape number " << i;
+template <typename T>
+void Check(std::initializer_list<int32_t> input1_shape, std::initializer_list<int32_t> input2_shape,
+           std::initializer_list<int32_t> output_shape, std::initializer_list<float> input1_data,
+           std::initializer_list<float> input2_data, std::initializer_list<float> output_data)
+{
+  float kQuantizedTolerance = GetTolerance(getMinValue(input1_data) + getMinValue(input2_data),
+                                           getMaxValue(input1_data) + getMaxValue(input2_data));
+  std::pair<float, int32_t> input1_quant_param =
+      quantizationParams<T>(getMinValue(input1_data), getMaxValue(input1_data));
+  std::pair<float, int32_t> input2_quant_param =
+      quantizationParams<T>(getMinValue(input2_data), getMaxValue(input2_data));
```

Comment:
Please move the quantization-related codes to the proper place. E.g., these lines do not have to be called if T is float.

Reply:
Oh, I found that In TFlite test case, it split uint8(or int8) case and float case for some reason. so, I just add a new test case for uint8 case.